### ISSUE TITLE
OAuth scopes by repository (or organization)

### ISSUE THREAD

**Issue by ePaul** [isaacs/github#731]
When an application using the API requests permissions from me and wants be able to do stuff with one repository, it has to request access to *all* repositories I have access to (whether I'm the owner or just member of an organization related to it).

E.g. Zappr (a PR approval tool) [currently needs read + write access to all repositories, and admin-access to the hooks](https://zappr.readthedocs.io/en/latest/faq/), though I only want to configure it for one repository. (Related: https://github.com/zalando/zappr/issues/90)

The OAuth flow should allow to request (and give) access for just one specific repository, instead all of them.

(Same is valid for organization-level permissions.)

**Comment by aspiers** [comment]
Yes, this is a massive pain, which prevents usage of OAuth apps in one organisation you're a member of when you can't or don't want to get approval from all the organisations you're in to give access to the app in question.  Here's an example of the problem:

![github-oauth2](https://user-images.githubusercontent.com/100738/41408020-42509546-6fc9-11e8-9894-b452e2681f26.png)

Here I want to test out GerritHub on a few repositories before potentially recommending that one or more organisations I'm in should adopt it.  But it wants to automatically grant access of the OAuth app to all organisations which haven't enabled third-party application access restrictions.  For the organisations I'm an admin of, I could enable this restriction, e.g.:

![3rd-party](https://user-images.githubusercontent.com/100738/41408253-efd5a490-6fc9-11e8-95d3-fae372f5cd7e.png)

but for the other organisations which don't yet have the restriction enabled, I would have to ask the admins of those organisations to enable third-party access restrictions before I could safely authorise the app without giving it unwanted access.

**Comment by aspiers** [comment]
See also https://github.com/dear-github/dear-github/issues/113

**Comment by lucamilanesio** [comment]
@aspiers thanks for raising this. We are planning to introduce an "Individual Maintainer" scope that would NOT require read:org. However, some of the GerritHub features won't be available, like the resolution of GitHub Teams in Gerrit ACLs.

It would be **great** if GitHub could allow to choose which organisations to share with GerritHub :-)

**Comment by aspiers** [comment]
According to https://github.com/houndci/hound/issues/925#issuecomment-398126761 apparently moving to GitHub Apps would allow per repo specific permissions.  Would this be another option for GerritHub?  (Although I appreciate it would presumably be a much bigger change.)

**Comment by aspiers** [comment]
[**@lucamilanesio**](https://github.com/lucamilanesio) commented on [14 Jun 2018, 21:40 BST](https://github.com/isaacs/github/issues/731#issuecomment-397431258 "2018-06-14T20:40:39Z - Replied by Github Reply Comments"):
> [@aspiers](https://github.com/aspiers) thanks for raising this. We are planning to introduce an "Individual Maintainer" scope that would NOT require read:org. However, some of the GerritHub features won't be available, like the resolution of GitHub Teams in Gerrit ACLs.

Looks like this has been done: [GerritHub.io and the new Reviewer role | GerritForge Blog](https://gitenterprise.me/2018/06/21/gerrithub-io-and-the-new-reviewer-role/)

**Comment by lucamilanesio** [comment]
@aspiers yes, and still allows to push changes to personal public projects
